psi4/src/psi4/libmints/writer.cc

```diff
@@ -930,42 +930,40 @@ void NBOWriter::write(const std::string &filename) {
     printer->Printf("\n $END");
 
     // Alpha AO->MO transformation
-    SharedMatrix soalphac = wavefunction_->Ca();
-    const Dimension aos = helper.petite_list()->AO_basisdim();
-    const Dimension nmo = wavefunction_->Ca()->colspi();
-    auto alphac = std::make_shared<Matrix>("Ca AO x MO", aos, nmo);
-    alphac->gemm(true, false, 1.00, sotoao, soalphac, 0.00);
+    SharedMatrix alphac = wavefunction_->Ca_subset("AO", "ALL");
 
     printer->Printf("\n $LCAOMO");
+
+    // Now write out the coefficients. NBO expects all coefficients
+    // for one MO before writing coefficients for the next MO.
+
     if (wavefunction_->same_a_b_orbs()) {
         for (int i = 0; i < nbf; i++) {
             for (int j = 0; j < nbf; j++) {
```

Comment:
This code assumes there are no linearly dependent functions!

Reply:
Do you have an example offhand? I can try to find one if not.

Reply:
Pretty easy to find, just pick a molecule and a large, augmented basis set. Here's an example
```
molecule {
0 1
H          0.62810        0.62810        0.62810
C          0.00000        0.00000        0.00000
H         -0.62810       -0.62810        0.62810
H         -0.62810        0.62810       -0.62810
H          0.62810       -0.62810       -0.62810
}

set basis aug-pcseg-3
set s_tolerance 1e-5
energy('scf')
```
output
```
  Minimum eigenvalue in the overlap matrix is 3.7150323933E-07.
  Using Canonical Orthogonalization with cutoff of 1.0000000000E-05.
  Overall, 3 of 281 possible MOs eliminated.
```

Reply:
For some reason, this is a **really** common bug...

Reply:
When I tried your example, Psi says NBO can't handle g basis functions, while QChem gave NBO the g basis functions _without_ eliminating linear dependencies. NBO handles the g functions and the dependencies just fine, eliminating the dependencies itself.

I'm inclined to let NBO sort out the linear dependencies itself, but teach Psi-NBO how to deal with g functions. Either way, I won't be getting this until tomorrow.

Reply:
You can remove the g functions from the basis, or compile a bigger version of Psi4. If you run in Q-Chem, you need to set the linear dependence threshold like in Psi4, since it's 1e-7 by default.

Currently, the code is just going to produce nonsense, since if you have linear dependencies, C is Nbf x Nmo, and so the last (Nbf-Nmo) columns of the matrix are going to run over the buffer and give you a segfault, or run and give garbage in the output. The linear dependence is **built in C**.

Reply:
The issue is pretty simple, just replace the column loop over nbf with a loop over the true number of columns.

Reply:
Thanks for the advice. Everything you said was clear after sleeping on it.

I tried the fix and am running into some problems with NBO syntax... I'll get back to this when the NBO devs get back to me.